File: QTVI-refactor/4_find_wave_bounds/SegmentPPG.cpp
#include "SegmentPPG.h"
#include "nanfastsmooth.h"
#include "StatsUtils.h"
#include "stdoutlier.h"
#include "RunLength.h"
#include <stdexcept>
#include <cmath>
#include <algorithm>

// Internal helper for segmentation logic
pair<vector<size_t>, vector<size_t>> segBeats(const vector<double>& ppg, const vector<int>& mask);
size_t newVallyFromPeaks(const vector<double>& ppg, size_t currVal, size_t currPeak,
    const vector<size_t>& peaks_idxs, const vector<size_t>& vally_idxs,
    const vector<bool>& peakoutliers, const vector<bool>& vallyoutlier);

SegmentPPGResult SegmentPPG(const vector<double>& ppg, double sampleRate) {
    SegmentPPGResult result;

    if (ppg.empty()) throw std::runtime_error("Empty signal");

    // Match MATLAB: ppg_smooth = nanfastsmooth(ppg, sampleRate * .25, 3);
    vector<double> ppg_smooth = nanfastsmooth(ppg, sampleRate * 0.25, 3);
    // Match MATLAB: M = movmean(ppg_smooth, sampleRate); 
    // Ensure movmean uses a centered window to match MATLAB's default
    vector<double> M = movmean(ppg_smooth, static_cast<size_t>(sampleRate));

    vector<int> p_mask(ppg.size(), 0);
    for (size_t i = 0; i < ppg.size(); ++i) {
        // Match MATLAB: p_mask(ppg_smooth > M) = 1
        p_mask[i] = (ppg_smooth[i] > M[i]) ? 1 : 0;
    }

    auto beatResult = segBeats(ppg, p_mask);
    vector<size_t> peakidx = beatResult.first;
    vector<size_t> vallyidx = beatResult.second;

    // CRITICAL: Replicate MATLAB crash/error behavior for flat/bad signals
    // MATLAB: if vallyidx(1) < peakidx(1) will throw if either is empty
    if (peakidx.empty() || vallyidx.empty()) {
        throw std::runtime_error("Incomplete beat segmentation (flat signal)");
    }

    if (std::abs(static_cast<int>(peakidx.size()) - static_cast<int>(vallyidx.size())) > 1) {
        throw std::runtime_error("not expecting this");
    }

    vector<bool> ppg_outliers_peaks = stdoutlier(
        vector<double>(peakidx.begin(), peakidx.end()), 2.5, 100, "both", false);
    vector<bool> vallyoutlier_time = stdoutlier(
        vector<double>(vallyidx.begin(), vallyidx.end()), 2.5, 100, "both", false);

    size_t pidx;
    // MATLAB is 1-based, C++ 0-based adjustment
    if (vallyidx[0] < peakidx[0]) {
        pidx = 0;
    }
    else {
        pidx = 1;
    }

    for (size_t vidx = 0; vidx < vallyidx.size() - 1; ++vidx) {
        if (vidx == 0 || vidx == vallyidx.size() - 2) {
            pidx++;
            continue;
        }

        if (vallyoutlier_time[vidx]) {
            vallyidx[vidx] = newVallyFromPeaks(ppg, vidx, pidx, peakidx, vallyidx,
                ppg_outliers_peaks, vallyoutlier_time);
        }
        pidx++;
    }

    result.ppgMinAmps = vallyidx;
    result.maxAmps = peakidx;
    return result;
}

pair<vector<size_t>, vector<size_t>> segBeats(const vector<double>& ppg, const vector<int>& mask) {
    vector<int> B;
    vector<double> N, BI;
    RunLength(mask, B, N, BI);

    vector<size_t> peaks, valleys;
    for (size_t i = 0; i < B.size(); ++i) {
        size_t start = static_cast<size_t>(BI[i]);
        size_t end = start + static_cast<size_t>(N[i]) - 1;
        if (end >= ppg.size()) end = ppg.size() - 1;

        if (B[i] == 0) { // Low region (Valleys)
            auto res = min_element_index(ppg, start, end + 1);
            valleys.push_back(start + res.second);
        }
        else { // High region (Peaks)
            auto res = max_element_index(ppg, start, end + 1);
            peaks.push_back(start + res.second);
        }
    }
    return { peaks, valleys };
}

size_t newVallyFromPeaks(const vector<double>& ppg, size_t currVal, size_t currPeak,
    const vector<size_t>& peaks_idxs, const vector<size_t>& vally_idxs,
    const vector<bool>& peakoutliers, const vector<bool>& vallyoutlier) {

    if (currPeak == 0 || currPeak >= peaks_idxs.size()) return vally_idxs[currVal];

    size_t start = peaks_idxs[currPeak - 1];
    size_t end = peaks_idxs[currPeak];
    auto res = min_element_index(ppg, start, end + 1);
    size_t new_idx = start + res.second;

    if (!peakoutliers[currPeak - 1] && !peakoutliers[currPeak]) {
        return new_idx;
    }
    return vally_idxs[currVal];
}

File: daniel_script/helpers/waves/ppg/SegmentPPG.m
function [ppgMinAmps, maxAmps] = SegmentPPG(ppg, sampleRate)

    %agressive smooth and tight moving mean
    ppg_smooth = nanfastsmooth(ppg, sampleRate * .25, 3);
    M = movmean(ppg_smooth, sampleRate);

    plow = ppg;
    plow(ppg_smooth > M) = nan;
    p_mask(~isnan(plow)) = 0;
    p_mask(isnan(plow)) = 1;
    [peakidx, vallyidx] = segBeats(ppg, p_mask);

    if abs(length(peakidx) - length(vallyidx)) > 1
        error('not expecting this')
    end

    ppg_outliers_peaks = stdoutlier(peakidx', 2.5, 100, 'both', 0);
    vallyoutlier_time = stdoutlier(vallyidx', 2.5, 100, 'both', 0);
    %vallyoutlier_amp = stdoutlier(ppg(vallyidx), 2.5, 100,'both', 0);

    if vallyidx(1) < peakidx(1)%vally peak vally peak ....
            pidx = 1;
    else % peak valley peak valley ....
            pidx = 2;
    end

    %
    %     close all;
    %     figure('units','normalized','outerposition',[0 0 1 1])
    %     plot(ppg,'Color',rgb('grey'));
    %     hold on;
    %     plot(ppg_smooth,':','Color',rgb('white'))
    %     plot(M,'Color',rgb('yellow'));
    %     plot(plow,'k--');
    %
    %     plot(peakidx, ppg(peakidx), '*w')
    %     plot(vallyidx, ppg(vallyidx), 'ow')
    %
    %     set(gca,'Color','k')

    for vidx = 1:length(vallyidx) - 1

        if vidx == 1 || vidx == length(vallyidx) - 1
            pidx = pidx + 1;
            continue
        end

        if vallyoutlier_time(vidx)
            %             v1 = vline(peakidx(pidx),'r');
            %             v2 = vline(vallyidx(vidx),'g');
            %             xlim([vallyidx(vidx)-1000,vallyidx(vidx)+1000]);
            vallyidx(vidx) = newVallyFromPeaks(ppg, vidx, pidx, peakidx, vallyidx, ppg_outliers_peaks, vallyoutlier_time);
            %             delete(v1);
            %             delete(v2);
        end

        pidx = pidx + 1;

    end

    %

    %   debug plot

    % %         vline(vallyidx(ppg_outliers_peaks),'--g');
    % %         vline(vallyidx(ppg_outliers_vallies),'--c');
    %
    %     h = [];
    %     h(end+1) = plot(nan,':','Color',rgb('white'));
    %     h(end+1) = plot(nan,'Color',rgb('yellow'));
    %
    %     legend(h,{'\color{white} Smooth PPG','\color{white} Moving Mean'});
    %     minOutliers = stdoutlier(vallyidx', 2.5, 10, 0);
    %     plot(vallyidx(minOutliers), ppg(vallyidx(minOutliers)), 'or')
    %
    %
    %
    %
    %     [B, N, BI] = RunLength(minOutliers);
    %     if ~isempty(N(N>3 & B == 1))
    %         idxs = find((N>3 & B == 1) == 1);
    %         for q = 1:length(idxs)
    %             idx1 = BI(idxs(q));
    %             idx2 = BI(idxs(q))+N(idxs(q))-1;
    %             section = vallyidx(idx1:idx2);
    %             out = PpgBestFitOfBeatTrain(vallyidx(1:idx1-1), vallyidx(idx2+1:end), section, 15);
    %             vallyidx = [vallyidx(1:idx1-1) out vallyidx(idx2+1:end)];
    %         end
    %     end

    ppgMinAmps = vallyidx;
    maxAmps = peakidx;
end

function [newidx] = newVallyFromPeaks(ppg, currVal, currPeak, peaks_idxs, vally_idxs, peakoutliers, vallyoutlier)
    beat = ppg(peaks_idxs(currPeak - 1):peaks_idxs(currPeak));
    [~, idx] = min(beat);

    if ~peakoutliers(currPeak - 1) &&~peakoutliers(currPeak)% no weird peaks

        if peaks_idxs(currPeak - 1) + idx - 1 == vally_idxs(currVal)
            newidx = vally_idxs(currVal);
        else
            newidx = peaks_idxs(currPeak - 1) + idx - 1;
        end

    else
        newidx = vally_idxs(currVal);
    end

end

function [dif] = oscillatingDiff(a, b)
    % diff skipping 1 each time.
    % assumtion is in time the two vectors would go [a,b,a,b,a,b,a,b...]
    % function gives back each ab pair diff, aka the diff on each odd index
    all = [a; b];
    all = sort(all);

    d = diff(all);
    dif = d(1:2:end);
end

function [peaks, vallies] = segBeats(ppg, mask)
    % find where 0 vs 1 and length of peroid
    [B(:, 1), B(:, 2), B(:, 3)] = RunLength(mask);

    % elements above and below line
    high = zeros(sum(B(:, 1)), 2);
    low = zeros(size(B, 1) - sum(B(:, 1)), 2);
    highidx = 1;
    lowidx = 1;

    for i = 1:size(B, 1)

        if B(i, 1) == 0
            low(lowidx, 1) = B(i, 3);
            low(lowidx, 2) = B(i, 3) + B(i, 2) - 1;
            lowidx = lowidx + 1;
        else
            high(highidx, 1) = B(i, 3);
            high(highidx, 2) = B(i, 3) + B(i, 2) - 1;
            highidx = highidx + 1;
        end

    end

    % find max
    peaks = zeros(size(high, 1), 1);

    for i = 1:size(high, 1)
        % max idx is indexed to begining of section (high(i,1)) so add that
        % to temp to move to proper position.
        [~, tmp] = max(ppg(high(i, 1):high(i, 2)));
        peaks(i) = tmp + high(i, 1);
    end

    peaks = peaks - 1;

    vallies = zeros(size(low, 1), 1);

    for i = 1:size(low, 1)
        % max idx is indexed to begining of section (high(i,1)) so add that
        % to temp to move to proper position.
        [~, tmp] = min(ppg(low(i, 1):low(i, 2)));
        vallies(i) = tmp + low(i, 1);
    end

    vallies = vallies - 1;

end

function [out] = PpgBestFitOfBeatTrain(sectionBefore, sectionAfter, dataToMatch, beats)

    try
        sectionBefore = sectionBefore(end - beats + 1:end);
    catch
    end

    try
        sectionAfter = sectionAfter(1:beats);
    catch
    end

    out = [];

    for i = 1:length(dataToMatch)
        medBefore = round(mean(diff(sectionBefore)));
        medAfter = round(mean(diff(sectionAfter)));

        endBefore = sectionBefore(end);
        beginAfter = sectionAfter(1);

        beforeRate = endBefore:medBefore:beginAfter;
        afterRate = endBefore:medAfter:beginAfter;

        if length(beforeRate) > length(afterRate)
            midLengths = round((beforeRate(1:numel(afterRate)) + afterRate) / 2);
        else
            midLengths = round((afterRate(1:numel(beforeRate)) + beforeRate) / 2);
        end

        v1 = vline(midLengths, 'r');

        if length(midLengths) < 2
            mid = endBefore + medBefore;
        else
            mid = midLengths(2);
        end

        errors = zeros(length(dataToMatch), 1);

        for z = 1:length(dataToMatch)
            errors(z) = abs(mid - dataToMatch(z));
        end

        [val, minIdx] = min(errors);
        out(end + 1) = dataToMatch(minIdx);

        sectionBefore(end + 1) = dataToMatch(minIdx);
        dataToMatch(dataToMatch <= minIdx) = [];

        delete(v1);
    end

end
